pdns/validate.cc

```diff
@@ -49,24 +49,26 @@ dState getDenial(const cspmap_t &validrrsets, const DNSName& qname, const uint16
         if(!nsec)
           continue;
 
-        /* check if the type is denied */
-        if(qname == v.first.first && !nsec->d_set.count(qtype)) {
-          LOG("Denies existence of type "<<QType(qtype).getName()<<endl);
-          return NXQTYPE;
-        }
-
         /* RFC 6840 section 4.1 "Clarifications on Nonexistence Proofs":
            Ancestor delegation NSEC or NSEC3 RRs MUST NOT be used to assume
            nonexistence of any RRs below that zone cut, which include all RRs at
            that (original) owner name other than DS RRs, and all RRs below that
            owner name regardless of type.
         */
-        LOG("type is "<<QType(qtype).getName()<<", NS is "<<std::to_string(nsec->d_set.count(QType::NS))<<", SOA is "<<std::to_string(nsec->d_set.count(QType::SOA))<<", signer is "<<getSigner(v.second.signatures).toString()<<", owner name is "<<v.first.first.toString()<<endl);
-        if (qtype != QType::DS && nsec->d_set.count(QType::NS) && !nsec->d_set.count(QType::SOA) &&
+        if (nsec->d_set.count(QType::NS) && !nsec->d_set.count(QType::SOA) &&
```

Comment:
The SOA check feels redundant, isn't the label count right after it sufficient?

Reply:
I think you are right, but since section 4.1 [1] of rfc6840 explicitly lists that condition and the check is cheaper than the label count one, I'd rather keep it. 

[1]: https://tools.ietf.org/html/rfc6840#section-4.1